lib/seahorn/BmcPass.cc

```diff
@@ -185,95 +196,188 @@ class BmcPass : public llvm::ModulePass {
-      bmc.addCutPoint(*dst);
-      LOG("bmc", errs() << "BMC from: " << src.bb().getName() << " to "
-                        << dst->bb().getName() << "\n";);
-
-      Stats::resume("BMC");
-      bmc.encode();
-
-      Stats::uset("bmc.dag_sz", dagSize(bmc.getFormula()));
-      Stats::uset("bmc.circ_sz", boolop::circSize(bmc.getFormula()));
-
-      LOG("bmc.simplify",
-          // --
-          Expr vc = mknary<AND>(bmc.getFormula());
-          Expr vc_simpl = z3_simplify(bmc.zctx(), vc);
-          llvm::errs() << "VC:\n"
-                       << z3_to_smtlib(bmc.zctx(), vc) << "\n~~~~\n"
-                       << "Simplified VC:\n"
-                       << z3_to_smtlib(bmc.zctx(), vc_simpl) << "\n");
-
-      if (m_out)
-        bmc.toSmtLib(*m_out);
+        Stats::resume("BMC.solve");
+        auto res = bmc.solve();
+        Stats::stop("BMC.solve");
 
-      if (!m_solve) {
-        LOG("bmc", errs() << "Stopping before solving\n";);
         Stats::stop("BMC");
-        return false;
-      }
 
-      Stats::resume("BMC.solve");
-      auto res = bmc.solve();
-      Stats::stop("BMC.solve");
-
-      Stats::stop("BMC");
-
-      if (res)
-        outs() << "sat";
-      else if (!res)
-        outs() << "unsat";
-      else
-        outs() << "unknown";
-      outs() << "\n";
+        if (res)
+          outs() << "sat";
+        else if (!res)
+          outs() << "unsat";
+        else
+          outs() << "unknown";
+        outs() << "\n";
+
+        if (res)
+          Stats::sset("Result", "FALSE");
+        else if (!res)
+          Stats::sset("Result", "TRUE");
+
+        LOG(
+            "bmc_core",
+            // producing bmc core is expensive. Enable only if specifically
+            // requested
+            if (!res) {
+              ExprVector core;
+              bmc.unsatCore(core);
+              errs() << "CORE BEGIN\n";
+              for (auto c : core)
+                errs() << *c << "\n";
+              errs() << "CORE END\n";
+            });
+
+        LOG(
+            "cex", if (res) {
+              errs() << "Analyzed Function:\n" << F << "\n";
+              errs() << "Trace \n";
+              BmcTrace trace(bmc.getTrace());
+              trace.print(errs());
+            });
+
+        if (res) {
+          StringRef CexFileRef(HornCexFile);
+          if (CexFileRef != "") {
+            if (CexFileRef.endswith(".ll") || CexFileRef.endswith(".bc")) {
+              auto &tli = getAnalysis<TargetLibraryInfoWrapperPass>();
+              auto const &dl = F.getParent()->getDataLayout();
+              BmcTrace trace(bmc.getTrace());
+              BmcTraceWrapper trace_wrapper(trace);
+              dumpLLVMCex(trace_wrapper, CexFileRef, dl, tli.getTLI(F),
+                          F.getContext());
+            } else {
+              WARN << "The Bmc engine only generates harnesses in bitcode "
+                      "format";
+            }
+          }
+        }
+      } else {
```

Comment:
break into functions. This is getting a little excessive.

Reply:
Not sure how to break into functions since BmcEngine <-> SolverBmcEngine and BmcTrace <-> SolverBmcTrace has diverged quite a bit; perhaps it will better to unite BmcEngine, SolverBmcEngine and PathBmcEngine under a single template parent class?

Reply:
I mean just break the big method into functions. Do not try to refactor. Just push bodies of if-then-else into functions